Add persistence logic around allocated streams in Camera AVSM

pidarped opened the pull request:
#### Summary
* Add templatized Store and Load functions for allocated Video/Audio/Snapshot streams.
  -- Invoke Store functions on Add/Modify/Remove stream functions -- Invoke Load functions from initial Load persistent  
      attributes call in Init().
* Mark the corresponding stream in HAL as allocated after loading from persisted state.
  -- Consolidate delegate APIs around loading of allocated streams. -- Use the PersistentAttributesLoadedCallback() API to  
      signal the HAL to start corresponding loaded streams.


#### Testing
* chip-camera-app commissioned with camera-controller
* allocate streams from camera-controller, restart chip-camera-app and read allocated streams attribute.
* deallocate streams from camera-controller, restart chip-camera-app and read allocated streams attribute.

#### Readability checklist

The checklist below will help the reviewer finish PR review in time and keep the
code readable:

-   [ ] PR title is
        [descriptive](https://project-chip.github.io/connectedhomeip-doc/contributing/pull_request_guidelines.html#title-formatting)
-   [ ] Apply the
        [_“When in Rome…”_](https://project-chip.github.io/connectedhomeip-doc/style/CODING_STYLE_GUIDE.html)
        rule (coding style)
-   [ ] PR size is short
-   [ ] Try to avoid "squashing" and "force-update" in commit history
-   [ ] CI time didn't increase

See: [Pull Request Guidelines](https://project-chip.github.io/connectedhomeip-doc/contributing/pull_request_guidelines.html)

pidarped committed:
Add persistence logic around allocated streams in Camera AVSM

* Add templatized Store and Load functions for allocated Video/Audio/Snapshot streams.

  -- Invoke Store functions on Add/Modify/Remove stream functions
  -- Invoke Load functions from initial Load persistent attributes call in
     Init().

* Mark the corresponding stream in HAL as allocated after loading from persisted state.

  -- Consolidate delegate APIs around loading of allocated streams.
  -- Use the PersistentAttributesLoadedCallback() API to signal the HAL to
     start corresponding loaded streams.

pidarped committed:
Incorporate automated review comments.

andy31415 reviewed (dismissed):
Please do not use the `SafeAttributePersistenceProvider`. Use `AttributePersistenceProvider`

pidarped committed:
Use AttributePersistenceProvider for the stream Load/Store APIs.

bzbarsky-apple commented on src/lib/support/CodeUtils.h:
```diff
@@ -257,7 +257,7 @@
         {                                                                                                                          \
             ChipLogError(MOD, MSG ": %" CHIP_ERROR_FORMAT, ##__VA_ARGS__, __err.Format());                                         \
-            return _err;                                                                                                           \
+            return __err;                                                                                                          \
```
While you're here, want to fix the typo in the "brief" bit?  Should be "log" instead of "lg".

pidarped committed:
Add fault injection based test to validate Store/Load for video streams.

pidarped committed:
Address review comments.

-Modify template function to only use the passed AttributeId to derive the
 streams vector and stream type.

pidarped commented on examples/camera-app/linux/include/camera-device.h:
```diff
@@ -65,7 +65,7 @@ static constexpr uint16_t kMaxImageRotation          = 359; // Spec constraint
 static constexpr uint8_t kMaxUserDefinedZones        = 10;  // Spec has min 5
 static constexpr uint8_t kSensitivityMax             = 10;  // Spec has 2 to 10
-
+static constexpr uint16_t kInvalidStreamID           = 65500;
```
The StreamIds start at 0 and monotonically increase. This was basically chosen as a large value. Modified to 0xFFFF and added a comment as well.

pidarped commented on src/app/clusters/camera-av-stream-management-server/camera-av-stream-management-server.cpp:
```diff
@@ -1639,6 +1701,79 @@ CHIP_ERROR CameraAVStreamMgmtServer::LoadStreamUsagePriorities()
+                                                           StreamType streamType)
+{
+    uint8_t buffer[N];
+    chip::TLV::TLVWriter writer;
```
Prefix removed.

pidarped commented on src/app/clusters/camera-av-stream-management-server/camera-av-stream-management-server.cpp:
```diff
@@ -1639,6 +1701,79 @@ CHIP_ERROR CameraAVStreamMgmtServer::LoadStreamUsagePriorities()
+    ReturnErrorOnFailure(writer.EndContainer(arrayType));
+
+    size_t len = writer.GetLengthWritten();
+    if (len == 0 && !streams.empty())
```
That's right. Removed.

pidarped committed:
Remove unused import in Test script.

pidarped committed:
Replace manual calculation with TLV::EstimateStructOverhead for
TLV encoded size calculation.

pidarped committed:
Fix deleted push-back to vector during AddStream.

pidarped committed:
Fix test scripts to perform a clean-up at the end to deallocate and
remove any persisted streams, allowing the test to be repeatable and
subsequent tests to be run without any remnant state.

pidarped committed:
Extract the stream persistence and attribute change notification into a
private helper method.

andy31415 reviewed (changes requested):
Changes requested: please add unit tests for serialize/deserialize logic at least. It should be trivially unit testable.

pidarped commented on src/app/clusters/camera-av-stream-management-server/camera-av-stream-management-server.h:
```diff
@@ -608,6 +651,12 @@ class CameraAVStreamMgmtServer : public CommandHandlerInterface, public Attribut
+    CHIP_ERROR PersistAndNotify();
+
+    template <AttributeId TAttributeId>
+    friend struct StreamTraits;
```
friend because the StreamTrait template accesses the private stream vector member of the cluster. This helper template struct helps in avoiding having to duplicate logic on all the 3 stream types around TLV encoding in the main Store/Load functions. Will add a comment above.

yufengwangca commented on src/python_testing/TC_AVSM_2_15.py:
```diff
@@ -168,6 +168,17 @@ async def test_TC_AVSM_2_15(self):
+        for stream in aAllocatedSnapshotStreams:
+            try:
+                await self.send_single_cmd(endpoint=endpoint, cmd=commands.SnapshotStreamDeallocate(snapshotStreamID=(stream.snapshotStreamID)))
+            except InteractionModelError as e:
```
We need to import 
"from matter.interaction_model import InteractionModelError, Status"

pidarped committed:
Add necessary import in TC_AVSM_2_15.py.

bzbarsky-apple commented on src/app/clusters/camera-av-stream-management-server/camera-av-stream-management-server.cpp:
```diff
@@ -1639,6 +1701,79 @@ CHIP_ERROR CameraAVStreamMgmtServer::LoadStreamUsagePriorities()
+    writer.Init(buffer, sizeof(buffer));
+
+    TLV::TLVType arrayType;
+    ReturnErrorOnFailure(writer.StartContainer(TLV::AnonymousTag(), TLV::kTLVType_Array, arrayType));
```
If we are sure this is always just going to be a vector of structs and nothing else needs to be stored, sure.  If we want more flexibility, we should wrap another TLV struct around the whole thing.

pidarped committed:
Address review comments

andy31415 reviewed (approved):
Switching my review to an approve: will defer modularity for this to happen together with code-driven migration (which will simplify unit testing even more). I was assured this will happen in the near future rather than being dropped.
